core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -111,48 +122,58 @@ private[kafka] object LogValidator extends Logging {
                                          compactedTopic: Boolean,
                                          timestampType: TimestampType,
                                          timestampDiffMaxMs: Long,
                                          partitionLeaderEpoch: Int): ValidationAndOffsetAssignResult = {
     var maxTimestamp = RecordBatch.NO_TIMESTAMP
     var offsetOfMaxTimestamp = -1L
     val initialOffset = offsetCounter.value
+    var isMagicV2 = false
 
     for (batch <- records.batches.asScala) {
-      ensureNonTransactional(batch)
+      validateBatch(batch)
+
+      var maxBatchTimestamp = RecordBatch.NO_TIMESTAMP
+      var offsetOfMaxBatchTimestamp = -1L
 
       for (record <- batch.asScala) {
-        record.ensureValid()
-        ensureNotControlRecord(record)
-        validateKey(record, compactedTopic)
+        validateRecord(batch, record, now, timestampType, timestampDiffMaxMs, compactedTopic)
 
         val offset = offsetCounter.getAndIncrement()
-        if (batch.magic > RecordBatch.MAGIC_VALUE_V0) {
-          validateTimestamp(batch, record, currentTimestamp, timestampType, timestampDiffMaxMs)
-
-          if (record.timestamp > maxTimestamp) {
-            maxTimestamp = record.timestamp
-            offsetOfMaxTimestamp = offset
-          }
+        if (batch.magic > RecordBatch.MAGIC_VALUE_V0 && record.timestamp > maxBatchTimestamp) {
+          maxBatchTimestamp = record.timestamp
+          offsetOfMaxBatchTimestamp = offset
         }
       }
 
+      if (batch.magic > RecordBatch.MAGIC_VALUE_V0 && maxBatchTimestamp > maxTimestamp) {
+        maxTimestamp = maxBatchTimestamp
+        offsetOfMaxTimestamp = offsetOfMaxBatchTimestamp
+      }
+
       batch.setLastOffset(offsetCounter.value - 1)
 
-      if(batch.magic >= RecordBatch.MAGIC_VALUE_V2)
+      if (batch.magic >= RecordBatch.MAGIC_VALUE_V2)
         batch.setPartitionLeaderEpoch(partitionLeaderEpoch)
 
-      // TODO: in the compressed path, we ensure that the batch max timestamp is correct.
-      //       We should either do the same or (better) let those two paths converge.
-      if (batch.magic > RecordBatch.MAGIC_VALUE_V0 && timestampType == TimestampType.LOG_APPEND_TIME)
-        batch.setMaxTimestamp(TimestampType.LOG_APPEND_TIME, currentTimestamp)
+      if (batch.magic > RecordBatch.MAGIC_VALUE_V0) {
+        if (timestampType == TimestampType.LOG_APPEND_TIME)
+          batch.setMaxTimestamp(TimestampType.LOG_APPEND_TIME, now)
+        else
+          batch.setMaxTimestamp(timestampType, maxBatchTimestamp)
+      }
+
+      isMagicV2 = batch.magic >= RecordBatch.MAGIC_VALUE_V2
     }
 
     if (timestampType == TimestampType.LOG_APPEND_TIME) {
-      maxTimestamp = currentTimestamp
-      offsetOfMaxTimestamp = initialOffset
+      maxTimestamp = now
+      if (isMagicV2)
+        offsetOfMaxTimestamp = offsetCounter.value - 1
+      else
+        offsetOfMaxTimestamp = initialOffset
```

Comment:
I thought that the message format version 2 should behave the same as the compressed case because we are dealing with record batches, so it doesn't make sense to return an offset of a record inside the batch. Is that wrong? If so, the implementation of `MemoryRecordsBuilder.info` is also incorrect.